doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,302 @@
+   their names will be suffixed with `reconfig-`. They will move to active
+   once the reconfiguration is finalized. These will be stored in the
+   `mz_pending_cluster_replicas` table.
+ - Reconfiguring Cluster: any cluster with one or more replicas in the
+   `mz_pending_cluster_replicas` table
+
+### Guard Rails
+- Only one reconfiguration will occur at a time for a given cluster. Alter statements
+  run against a reconfiguring cluster will fail.
+- We must protect against creating a source/sink on a reconfiguring cluster and
+  reconfiguring a cluster with sources/sinks. These clusters will need to use
+  the non-scheduled mechanism (delete-before-create). This is due to a limitation
+  where sources and sinks must only run on a single replica.
+- We must store all state required to restart a `ReconfigureCluster` task in the event that environmentd
+  crashes/restarts.
+- We may want to consider limiting timeout duration.
+
+### Visibility
+Reconfiguration replicas should be identifiable for purposes of reporting and
+visibility in the UI. We may be able to do this by creating a prefix `reconfig-`
+for reconfiguration replicas, alternatively, we could set a bool status on the
+replica resource in the catalog.
+
+### Details
+
+#### Limitations:
+  - Cannot be used on clusters with sources, sources should not be created on
+    clusters with reconfiguration replicas.
+  - Must be run on managed clusters, and cannot be used to alter a cluster from
+    managed to unmanaged.
+
+#### Replica Naming
+Active replicas will continue with the same naming scheme (r1, r2, ...)
+Reconfiguration replicas will use the `reconfig` prefix; ex `reconfig-r3`,
+`reconfig-r4`, which will be removed once they are moved to active.
+
+#### Interactions with other schedules:
+
+__V1__
+We will choose to initially disallow any reconfiguration of a cluster
+with a schedule.
+
+__V2__
+For refresh schedules, we will only allow graceful reconfiguration on clusters
+whose refresh schedule is "ON", we would also need to prevent the clusters from being
+turned off until the reconfiguration is finished.
+
+
+__Future Schedules__
+Some proposed schedules, auto-scaling, for instance, may want to use graceful
+reconfiguration under the hood. These would need specific well-defined
+interactions with user-invoked graceful reconfiguration.
+
+
+#### Catalog Changes
+We will need to create a new table `mz_internal.mz_pending_cluster_replicas`
+
+| column     |  type |
+| ---------- | ----- |
+| replica_id | str   |
+| cluster_id | str   |
+| timeout    | int   |
+| wait_type  | str   |
+| background | bool  |
+| start_time | int   |
+| finished   | bool  |
+| continue_on_timeout | bool |
+
+This table will be responsible for holding the state of the graceful replication task.
+Each reconfiguration replica will have an entry in `mz_epnding_cluster_replicas`.
+
+
+#### Reconfiguration Task
+
+The reconfiguration task will perform the following actions:
```

Comment:
What is the task exactly? A tokio task? A coordinator message that gets sequenced and can possibly send other coord messages?

Reply:
Addressed in new update, and discussed in huddle, but I'll do a quick outline here.

Initially the thought was a tokio task, however, it seems like it'll be cleaner to do this through a cluster pending reconfiguration message handler that performs the check, finalizes the reconfiguration and returns to the user.

We'll have to adjust the sequencer to know it shouldn't retire the connection if it's a graceful reconfiguration.